File: 79-word-search/word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int n = board.length;
        int m = board[0].length;

        // Iterate over the board to find the starting character
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == word.charAt(0) && backtrack(board, word, i, j, 0)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean backtrack(char[][] board, String word, int i, int j, int index) {
        int n = board.length;
        int m = board[0].length;

        // If the entire word is found
        if (index == word.length()) {
            return true;
        }

        // Check boundaries and character match
        if (i < 0 || i >= n || j < 0 || j >= m || board[i][j] != word.charAt(index)) {
            return false;
        }

        // Temporarily mark the cell as visited
        char temp = board[i][j];
        board[i][j] = '#';

        // Explore all 4 possible directions
        boolean found = 
            backtrack(board, word, i, j + 1, index + 1) || // Right
            backtrack(board, word, i + 1, j, index + 1) || // Down
            backtrack(board, word, i - 1, j, index + 1) || // Up
            backtrack(board, word, i, j - 1, index + 1);   // Left

        // Restore the cell's original value
        board[i][j] = temp;

        return found;
    }
}

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:

        n = len(board)
        m = len(board[0])

        def backtrack(i,j,index):

            if index == len(word):
                return True

            if i<0 or i >= n or j < 0 or j >= m or board[i][j] != word[index]:
                return False


            temp = board[i][j]
            board[i][j] = '#'

            found = backtrack(i,j+1,index+1) or backtrack(i+1,j,index+1) or backtrack(i-1,j,index+1) or backtrack(i,j-1,index+1)

            board[i][j] = temp
            return found

        for i in range(n):
            for j in range(m):
                if board[i][j] == word[0] and backtrack(i,j,0):
                    return True
        return False


